relay-sampling/src/evaluation.rs

```diff
@@ -24,13 +28,116 @@ fn pseudo_random_from_uuid(id: Uuid) -> f64 {
     generator.sample(dist)
 }
 
+/// Increments the reservoir count for a given rule in redis.
+///
+/// - INCR docs: [`https://redis.io/commands/incr/`]
+/// - If the counter doesn't exist in redis, a new one will be inserted.
+#[cfg(feature = "redis")]
+fn increment_redis_reservoir_count(
+    redis: &Arc<RedisPool>,
+    org_id: u64,
+    rule_id: RuleId,
+) -> anyhow::Result<i64> {
+    let mut command = relay_redis::redis::cmd("INCR");
+
+    let key = format!("bias:{}:{}", org_id, rule_id);
+    command.arg(key);
+
+    Ok(command.query(&mut redis.client()?.connection()?)?)
+}
+
+/// The amount of matches for each reservoir rule in a given project.
+pub type ReservoirCounters = Arc<Mutex<BTreeMap<RuleId, i64>>>;
+
+/// Utility for evaluating reservoir-based sampling rules.
+///
+/// A "reservoir limit" rule samples every match until its limit is reached, after which
+/// the rule is disabled.
+///
+/// This utility uses a dual-counter system for enforcing this limit:
+///
+/// - Local Counter: Each relay instance maintains a local counter to track sampled events.
+///
+/// - Redis Counter: For processing relays, a Redis-based counter provides synchronization
+///   across multiple relay-instances. When incremented, the Redis counter returns the current global
+///   count for the given rule, which is then used to update the local counter.
+#[derive(Debug)]
+pub struct ReservoirEvaluator {
+    counters: ReservoirCounters,
+    #[cfg(feature = "redis")]
+    redis_pool: Option<Arc<RedisPool>>,
+    #[cfg(feature = "redis")]
+    org_id: Option<u64>,
```

Comment:
Ideally this would be one option, since we only care about having both or none. We always have access to the org id but I opted for putting it behind an option rather than initializing with an invalid value, even though it wouldn't matter for the code execution